File: src/main/java/algo/breadthFirstSearch/RottingOranges_994.java
package algo.breadthFirstSearch;

import java.util.ArrayDeque;
import java.util.ArrayList;

public class RottingOranges_994 {
  public static int orangesRotting(int[][] grid) {
    int rows = grid.length, cols = grid[0].length;
    var q = new ArrayDeque<int[]>();
    for (var i = 0; i < rows; i++) {
      for (var j = 0; j < cols; j++) {
        if (grid[i][j] == 2) {
          q.offer(new int[]{i, j});
        }
      }
    }

    var minutes = 0;
    while (!q.isEmpty()) {
      var size = q.size();
      var rottings = new ArrayList<int[]>();
      for (var i = 0; i < size; i++) {
        var rotting = q.poll();
        int row = rotting[0], col = rotting[1];
        if (row + 1 < rows && grid[row + 1][col] == 1) {
          grid[row + 1][col] = 2;
          rottings.add(new int[]{row + 1, col});
        }
        if (row - 1 >= 0 && grid[row - 1][col] == 1) {
          grid[row - 1][col] = 2;
          rottings.add(new int[]{row - 1, col});
        }
        if (col + 1 < cols && grid[row][col + 1] == 1) {
          grid[row][col + 1] = 2;
          rottings.add(new int[]{row, col + 1});
        }
        if (col - 1 >= 0 && grid[row][col - 1] == 1) {
          grid[row][col - 1] = 2;
          rottings.add(new int[]{row, col - 1});
        }
      }
      rottings.forEach(q::offer);
      minutes++;
    }

    for (var r : grid) {
      for (var c : r) {
        if (c == 1) {
          return -1;
        }
      }
    }
    return Math.max(minutes - 1, 0);
  }

  public static void main(String[] args) {
    System.out.println(orangesRotting(new int[][]{{2,1,1},{1,1,0},{0,1,1}}));
    System.out.println(orangesRotting(new int[][]{{2,2},{1,1},{0,0},{2,0}}));
  }
}

File: src/main/kotlin/algo/breadthFirstSearch/RottingOranges_994.kt
package algo.breadthFirstSearch

import java.util.LinkedList

class KtRottingOranges_994 {
    companion object {
        const val FRESH = 1
        const val ROTTEN = 2
    }
    fun orangesRotting(grid: Array<IntArray>): Int {
        fun bfs(grid: Array<IntArray>): Int {
            var time = 0
            val visited = Array(grid.size) { BooleanArray(grid[0].size) }
            val q = LinkedList<IntArray>()
            for (row in grid.indices) {
                for (col in grid[0].indices) {
                    // init rotten oranges
                    if (grid[row][col] == ROTTEN) {
                        q.offer(intArrayOf(row, col))
                        visited[row][col] = true
                    }
                }
            }
            while (q.isNotEmpty()) {
                val size = q.size
                for (i in 1..size) {
                    val cell = q.poll()
                    val (m, n) = cell
                    grid[m][n] = ROTTEN
                    if (m + 1 < grid.size && grid[m + 1][n] == FRESH && !visited[m + 1][n]) {
                        q.offer(intArrayOf(m + 1, n))
                        visited[m + 1][n] = true
                    }
                    if (m - 1 >= 0 && grid[m - 1][n] == FRESH && !visited[m - 1][n]) {
                        q.offer(intArrayOf(m - 1, n))
                        visited[m - 1][n] = true
                    }
                    if (n + 1 < grid[0].size && grid[m][n + 1] == FRESH && !visited[m][n + 1]) {
                        q.offer(intArrayOf(m, n + 1))
                        visited[m][n + 1] = true
                    }
                    if (n - 1 >= 0 && grid[m][n - 1] == FRESH && !visited[m][n - 1]) {
                        q.offer(intArrayOf(m, n - 1))
                        visited[m][n - 1] = true
                    }
                }
                time++
            }
            return (time - 1).coerceAtLeast(0)
        }

        val time = bfs(grid)
        for (row in grid.indices) {
            for (col in grid[0].indices) {
                // fresh orange found
                if (grid[row][col] == FRESH) {
                    return -1
                }
            }
        }
        return time
    }
}

fun main() {
    val rotting = KtRottingOranges_994()
    println(rotting.orangesRotting(arrayOf(intArrayOf(2,1,1), intArrayOf(1,1,0), intArrayOf(0,1,1))))
    println(rotting.orangesRotting(arrayOf(intArrayOf(2,2), intArrayOf(1,1), intArrayOf(0,0), intArrayOf(0,2))))
}